features/ALS012_Alignment-segment-start-and-length-attribute-types.feature

```diff
@@ -0,0 +1,25 @@
+@implementer-agreement
+@ALS
+@version1
+@E00010
+
+Feature: ALS012 - Alignment segment start and length attribute types
+
+  The rule verifies that an alignment segment uses the correct IfcLengthMeasure type
+  for attributes SegmentStart and SegmentLength.
+  Ref: [Issue #141](https://github.com/buildingSMART/ifc-gherkin-rules/issues/141)
+
+Scenario Outline: Correct entity type used for SegmentStart and Segment Length
+  Given an <entity>
+  Given its attribute Segments
+
+  Then The type of attribute <attribute> must be IfcLengthMeasure
+
+  Examples:
+    | entity                      | attribute     |
+    | IfcCompositeCurve           | SegmentStart  |
```

Comment:
I think this works because in "The type of attribute ..." impl we do `getattr(i, attribute, None)` so if the attribute does not exist it doesn't raise an exception and the `None`s are later filtered out.

But! It probably is necessary though to implement a bit more formally in the gherkin logic that this only (a) applies to alignment representations or (b) only involves segments of type IfcCurveSegment and not IfcCompositeCurveSegment (the most confusing naming distinction ever!! still upset about this 😠 ). Because what would happen otherwise is that even a non-suspecting IFC2X3 file can get green checkmarks on alignment related rules.

Reply:
I added a few givens to ensure green checks only for Segments that are part of an alignment representation in a 4.3 model.  I also added `without subtypes` to address @Ghesselink 's comments.

Reply:
```
  Given An IfcAlignment
  Given Its attribute Representation
  Given Its attribute Representations
  Given an <entity> without subtypes
```

Thanks, but this only partially solves the problem. Because with a new `Given an {entity}` there is no correspondence with the lines above. I haven't checked how it would work in practice, but I guess it would just overwrite the context, which still means that IfcCompositeCurves not part of an IfcAlignment would be subsequently processed (as long as there is at least a single IfcAlignment in the model, not even sure about that last part).

What I think we could do is use the ifcopenshell.file.traverse() method.

```
  Given An IfcAlignment
  Given Its attribute Representation
  Given Its attribute Representations
  Given All refererenced instances
  Given The instances with class name <entity>
```

`All refererenced instances` would do file.traverse() on the previous inst in the stack. `The instances with class name <entity>` would filter those based on entity name. That way we can basically arbitrarily jump in the graph to a location of interest. Or - what we're trying to solve here - maintain that there is a relationship between the first Given and the instance of type {entity} we're investigating.

Reply:
> What I think we could do is use the ifcopenshell.file.traverse() method.

Latest commit incorporates this into a step implementation.  I also added a second passing test file that contains an additional composite curve with segment start and end defined by `IfcParameterValue` used for solid extrusion of a column.  This second composite curve is correctly not flagged as an error since it is not part of an `IfcAlignment`.

Reply:
> Another note; aren't IfcSegmentedReferenceCurve and IfcGradientCurve subtypes of IfcCompositeCurve?

Excellent comment btw, but it is addressed in the new structure because `Its Entity Type` does not take into account inheritance because it takes the entity types as strings and pushes these to the next stack layer.

https://github.com/buildingSMART/ifc-gherkin-rules/blob/579e66b94b5bb7270c6dfa60b5c7a305b92dc9bc/features/steps/givens/entities.py#L52-L57